src/protocols/secure_channel/CASESession.h

```diff
@@ -146,7 +146,7 @@ class DLL_EXPORT CASESession : public Messaging::UnsolicitedMessageHandler,
      * @param session     Reference to the secure session that will be initialized once session establishment is complete
      * @return CHIP_ERROR The result of session derivation
      */
-    CHIP_ERROR DeriveSecureSession(CryptoContext & session) const override;
```

Comment:
Why do these various things become non-const?  Is this because PSA mutates some state of things it uses when doing key derivation?  Do the things it uses need to be members?  If so, would it be appropriate for them to be `mutable` members (much as I hesitate to suggest that)?

Reply:
It's because [psa_pake_get_shared_key()](https://github.com/ARM-software/psa-api/pull/114/files#diff-c5fda0d6f379353d5f5defd8e7e8fb3aec52e0d9806698104021dc7e18817f99R48) function takes a non-const pointer to the operation that is a member of `PSASpake2p_P256_SHA256_HKDF_HMAC` (which in turn is a member of `PASESession`). Turning the methods to non-const is my fault as I suggested @ArekBalysNordic that it would be a clear solution than `const_cast` or `mutable` :) (from my experience, some devs don't like `mutable` for things other than mutexes as it hides the fact of a method having a side effect but it's probably acceptable in this case).

Reply:
@bzbarsky-apple so it looks like we only need a non-const pointer for the PSA stuff, but not for the raw-key. We've got a few ways to tackle this: like I did before, by taking out the const from the "parent" methods, or we could try using `const_cast` or maybe even mutable. But As @Damian-Nordic wrote, it seemed like those options weren't really popular with the devs. What do you think? Is it a no-go to drop the `const` from those methods when we're working with the new Spake2+ stuff that can't be const?

Reply:
I don't think it's necessarily a problem; I am just trying to understand the situation.

So getting a shared key with PSA requires passing in some mutable state in addition to the place to put the key id?  And that mutable state needs to be in some way persistent (as in, the mutated version needs to be kept around)?  How long does that mutated state need to be kept around?

Reply:
That mutable state is `psa_pake_operation_t` that acts as a context for the entire PAKE operation:
1. First, you feed this context with subsequent input values like passwords, salt etc.
2. Then you use the same context to output the results. PSA APIs often require non-const pointers to the operation even for getting the results, probably to enable implementations to complete some calculations at the last moment.

Reply:
OK.  And I assume psa_pake_operation_t is non-copyable, in the sense that you couldn't have mutable methods that update its state with salt/etc and then a const method that copies the state, gets the key with the copied state (letting PSA do whatever computations on that copied state), and so on?

(Ignoring for the moment whether that would likely be inefficient; just trying to wrap my head around the problem space.)

Reply:
I can't find anything about copyability of the operation on the spec but given that there's no official API other than `memcpy` for that I doubt we can rely that it will always work. I guess if you think it's better not to remove the consts for this method just because PSA requires that, I would say adding `mutable` is the best workaround in this case.

Reply:
What do you think @bzbarsky-apple ?

Reply:
Let's just make these non-const, if there are crypto backends where this really is a conceptually non-const operation.